jetty-core/jetty-session/src/main/java/org/eclipse/jetty/session/AbstractSessionManager.java

```diff
@@ -1178,103 +1200,107 @@ protected RequestedSession resolveRequestedSessionId(Request request)
+                String name = getSessionIdPathParameterName();
+                String[] params = param.split(";");
+                for (String p : params)
                 {
-                    char c = param.charAt(i);
-                    if (c == ';' || c == '#' || c == '?' || c == '/')
-                        break;
-                    i++;
+                    if (!p.startsWith(name) || p.length() <= name.length() || p.charAt(name.length()) != '=')
+                        continue;
+
+                    if (ids == null)
+                        ids = new ArrayList<>();
+                    ids.add(p.substring(name.length() + 1).trim());
                 }
-                requestedSessionId = param.substring(s, i);
-                requestedSessionIdFromCookie = false;
+            }
+        }
 
-                if (LOG.isDebugEnabled())
-                    LOG.debug("Got Session ID {} from URL", requestedSessionId);
+        if (ids == null)
+            return NO_REQUESTED_SESSION;
+
+        if (LOG.isDebugEnabled())
+            LOG.debug("Got Session IDs {} from cookies {}", ids, cookieIds);
+
+        for (int i = 0; i < ids.size(); i++)
+        {
+            String id = ids.get(i);
+            if (session == null)
+            {
+                //we currently do not have a session selected, use this one if it is valid
+                ManagedSession s = getManagedSession(id);
+                if (s != null && s.isValid())
+                {
+                    //associate it with the request so its reference count is decremented as the
+                    //request exits
+                    requestedSessionId = id;
+                    requestedSessionIdFromCookie = i < cookieIds;
+                    session = s;
 
-                session = getManagedSession(requestedSessionId);
+                    if (LOG.isDebugEnabled())
+                        LOG.debug("Selected session {}", session);
+                }
+                else
+                {
+                    if (LOG.isDebugEnabled())
+                        LOG.debug("No session found for session id {}", id);
+
+                    //if we don't have a valid session id yet, just choose the first one
+                    if (requestedSessionId == null)
+                    {
+                        requestedSessionId = id;
+                        requestedSessionIdFromCookie = i < cookieIds;
+                    }
+                }
+            }
+            else
+            {
+                //we currently have a valid session selected. We will throw an error
+                //if there is a _different_ valid session id. Duplicate ids, or
+                //invalid session ids are ignored
+                if (!session.getId().equals(getSessionIdManager().getId(id)))
+                {
+                    if (LOG.isDebugEnabled())
+                        LOG.debug("Multiple different valid session ids: {}, {}", requestedSessionId, id);
```

Comment:
I think  you've got all the info to be able to output the source of the ids, isn't it something like:

LOG.debug("Multiple different valid session ids: {} from cookie {}, {} from cookie {}", requestedSessionId, requestedSessionIdFromCookie, id, i < cookieIds);

And similar for the other DEBUG log messages.